File: exercism/elixir/house/lib/house.ex
defmodule House do
  @verbs [
    "lay in",
    "ate",
    "killed",
    "worried",
    "tossed",
    "milked",
    "kissed",
    "married",
    "woke",
    "kept",
    "belonged to"
  ]

  @actors [
    "malt",
    "rat",
    "cat",
    "dog",
    "cow with the crumpled horn",
    "maiden all forlorn",
    "man all tattered and torn",
    "priest all shaven and shorn",
    "rooster that crowed in the morn",
    "farmer sowing his corn",
    "horse and the hound and the horn"
  ]

  @doc """
  Return verses of the nursery rhyme 'This is the House that Jack Built'.
  """
  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(start, stop) do
    all_phrases = phrases(1, stop, @actors, @verbs, []) |> Enum.reverse()

    start..stop
    |> Stream.map(fn verse_number ->
      verse =
        all_phrases
        |> Enum.slice(0, verse_number)
        |> Enum.reverse()
        |> Enum.join(" ")

      "This is #{verse}.\n"
    end)
    |> Enum.join()
  end

  defp phrases(1, stop, actors, verbs, _) do
    phrases(2, stop, actors, verbs, ["the house that Jack built"])
  end

  defp phrases(start, stop, [actor | actors], [verb | verbs], acc) when start <= stop do
    phrases(start + 1, stop, actors, verbs, ["the #{actor} that #{verb}" | acc])
  end

  defp phrases(_, _, _, _, acc), do: acc
end

File: exercism/lua/house/house.lua
local s = require("string")
    
local who = {
    "house that Jack built.", "malt", "rat", "cat",
    "dog", "cow with the crumpled horn", 
    "maiden all forlorn",
    "man all tattered and torn", 
    "priest all shaven and shorn",
    "rooster that crowed in the morn",
    "farmer sowing his corn",
    "horse and the hound and the horn"
    }

local verbs = {
    "lay in", "ate", "killed", "worried", "tossed",
    "milked", "kissed", "married", "woke", "kept",
    "belonged to"
}
    
local house = {}

house.verse = function(which)
    local res = s.format("This is the %s", who[which])
    for i=1, which - 1 do
        res = s.format("%s\nthat %s the %s", res, verbs[which - i], who[which - i])
    end
    return res
end

house.recite = function()
    local res = house.verse(1)
    for i=1, #who - 1 do
        res = s.format("%s\n%s", res, house.verse(i + 1))
    end
    return res
end

return house
